### PR TITLE
fix(warehouse-sources): keep a displaced staged cursor promotable

### PR DESCRIPTION
**Pull request by jakesciotto**
## Problem

A source that syncs newest-first never records a watermark when one sync run spans more than one extract attempt.
Every later run then re-reads the full history.
An append-only table receives every row again on each run, and billed rows grow with it.

How the watermark is lost, in the order to check it:

- The v3 pipeline stages the cursor under the key `{workflow_run_id}-a{attempt}`.
- The load processor promotes the cursor with the run uuid of the batch that completes.
- `stage_incremental_field_value` held one slot. A new attempt replaced that slot before the earlier batch landed.
- `promote_staged_incremental_values` then matched nothing. For a newest-first source (`sort_mode="desc"`), `finalize_desc_sort_incremental_value` is the only writer of the watermark, and the v3 path stages instead of writes.
- With no watermark, the source sends no incremental window. The next run is long again, and the loss repeats.

On one live source, the tables with short runs held a watermark and no staged cursor.
The tables with runs of many hours held no watermark and a staged cursor with `earliest_value` only.
That cursor moved to a new attempt while under observation.
A read of live schema state found the same signature on a handful of newest-first sources across several teams.
The full investigation is in #83855.

Refs #83855

## Changes

- A newest-first source now records its watermark when a run spans several attempts. The next run sends a narrower window, and an append table stops re-appending the whole history. A run that resumed mid-way promotes its final attempt's own maximum, so the next run re-reads the earlier attempts' rows once (see Not in this PR).
- A displaced staged cursor moves to `incremental_staged_pending` and stays promotable until its own batch lands. Promotion reads the live slot, or the parked list when a newer attempt displaced the run, and removes the entry it used. A run is in one of the two, never both.
- A run that stages again after a newer attempt displaced it continues its parked cursor. One run's `earliest_value` and `last_value` never split between the live slot and the parked list.
- Staging and promotion run under the schema's row lock through `update_sync_type_config_keys`. The outgoing attempt and the incoming attempt each hold a stale in-memory copy of the row, and an unlocked save from either copy erased the other's entry.
- The parked list holds at most `STAGED_CURSOR_PENDING_LIMIT` entries. That limit derives from the production retry cap in `retry_limits.py`, the same constant the import activity reads, so the two cannot drift. A resumable source gets that cap whatever its sync type.
- The parked list is not scoped to one run. Dead entries from abandoned attempts persist until the trim evicts them. The trim keeps the newest entries, and a live run's entries are always the newest, so a dead entry goes first.
- Promotion is monotonic. The high-water mark never moves backwards, and the low-water mark never moves forwards, because two attempts can now land out of order.
- For a datetime or date field, an epoch number converts to the type of its partner before the comparison, so an epoch cursor orders against an ISO string cursor.
- The load processor logs `staged_cursor_missing` when a completed incremental or append job has no staged cursor for its run uuid. That is the signature of this defect, and it was silent.
- `_finalize` stages the desc watermark before it sends the final-batch notification, so the load consumer cannot complete that batch and promote a slot the watermark has not reached yet.
- Mechanical: `update_sync_type_config_for_reset_pipeline` also clears the parked list. Nothing in the UI changes.

> [!NOTE]
> The monotonic guard falls back to "the newest promotion wins" when it cannot order a pair: a null `incremental_field_type`, a naive datetime against an aware one, or a millisecond epoch. ObjectID cursors of equal length order as strings.
> That fallback is the behavior every cursor type had before this PR, so no source changes behavior.
> Keeping the current value instead would freeze the watermark of every untyped schema for good.
> A read of live schema state found no schema whose promoted and staged cursors mix a naive and an aware datetime, and no millisecond epoch cursor.
> Ordering for those types is a follow-up.

The smallest fix was rejected.
Carrying a staged value forward when the run uuid changes can promote a value from an abandoned attempt.
That value can describe data that never landed, and a later promotion would skip that data for good.
A related proposal to keep resume state across scheduled jobs was rejected for the same reason.
This change keeps the existing rule.
A cursor is promoted only inside the completion transaction of its own batch.
No state survives across scheduled jobs.

> [!NOTE]
> A first pass over a history larger than one run still restarts at the newest record.
> `earliest_value` tracks how far back a run reached, and nothing consumes it yet.
> That is the second remedy in #83855. It waits until the promoted `earliest_value` after this fix shows whether a first pass completes in one run.
>
> Consequence for a source already in this state: once the watermark promotes, the source sends a `> last_value` window and stops walking history downward.
> The table keeps whatever depth the last full run reached, and it stops deepening.
> A team that needs the older history needs the second remedy, or a pipeline reset that re-reads everything.

Not in this PR:

- A resumed newest-first run promotes only its final attempt's maximum as `last_value`, because the in-memory maximum is seeded per attempt and resume state carries no cursor. The next run re-reads the earlier attempts' rows once. Follow-up: seed the in-memory maximum from resume state so the final attempt stages the run's true maximum.
- The field-change path in the schema API does not clear a staged or parked cursor. Unverified, and it gets its own PR.
- Ordering for naive datetime and millisecond epoch cursors in the monotonic guard.

## How did you test this code?

- New and changed cases in `TestStagedIncrementalCursor` in `test_models.py`. Each test name states the regression it catches. The groups: parking a displaced cursor, continuing a parked cursor for the same run, promotion from the parked list, both monotonic directions including an ObjectID pair, and a run uuid with no slot anywhere.
- `TestStagedIncrementalCursorStaleWriters` is database-backed. Two instances of one schema stage for two attempts of one run, each from a stale copy. It fails on an unlocked save, because the second writer erases the first writer's parked entry.
- `TestPromoteStagedCursor` in `test_processor.py` gains two cases. A completed incremental run with no staged cursor logs the warning, and a full-refresh run stays quiet.
- The epoch-against-string cases fail on the previous comparison code. That code treated a number and a datetime as unorderable, so an older epoch cursor moved the watermark backwards.
- Run locally against Postgres and ClickHouse: `test_models.py` and `test_processor.py` in full, and repo-wide `mypy --cache-fine-grained .` as CI runs it.

## Automatic notifications

- [ ] Publish to changelog?

## Docs update

None. No documented setting or workflow changes.

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Authored with Claude Code and PostHog Desktop. Skills invoked across the sessions: `/writing-pr-descriptions`, `/writing-tests`, `/writing-code-comments`, `/writing-simplified-technical-english`, `/security-review`.

Steps explored, in order:

1. The issue first read as "two tables never finish a sync". The job history showed that nearly every run completes, so the issue was restated as a watermark defect.
2. A comparison of the tables on one source showed the split follows run length, not sync type. Short runs held a watermark. Long runs held none.
3. The staged cursor's `run_uuid` advanced to a new attempt within minutes while the run was live. That pointed at attempt-scoped staging.
4. The code trace confirmed it. The source sends no incremental window without a watermark, the newest-first branch stages `earliest_value` only, the only newest-first writer stages in v3 instead of writes, and the single slot is replaced when the attempt advances.
5. Carrying the staged value forward across run uuids was rejected (see Changes). The fix changes when promotion happens, not how long stale state survives.
6. Parking with same-transaction promotion was chosen. The monotonic guard followed, because out-of-order landing became possible.
7. A review pass found nine defects in the first version. Six land here: the unlocked read-modify-write, the same-run split between the live slot and the parked list, the silent missing-cursor completion, the bound comment, the unparameterized types, and the duplicated filter. Three are listed under "Not in this PR" with the evidence that moved them.

No duplicate: `gh pr list --state open --search "staged cursor incremental watermark"` found no other PR on this defect.
Public artifact: the investigation drew on live sources' sync history. This description states those findings without figures, and the tests use invented values.

---
*Created with [PostHog Desktop](https://posthog.com/desktop?ref=pr)*

### CODE REVIEW

File: products/warehouse_sources/backend/models/external_data_schema.py [hunk 1]
```diff
@@ -832,16 +841,101 @@ def stage_incremental_field_value(self, run_uuid: str, last_value: Any, earliest
...
+            return None
+        if left is None or right is None:
+            return None
+        if isinstance(left, bool) or isinstance(right, bool):
+            return None
+        left_is_number = isinstance(left, int | float)
+        right_is_number = isinstance(right, int | float)
+        if left_is_number != right_is_number:
+            # An epoch cursor against a parsed datetime for the same field: not safely orderable.
+            return None
```
[the remark this answers is not in the document]
Comment (jakesciotto) [reply products/warehouse_sources/backend/models/external_data_schema.py]: Fixed in 9a99c0d. `_align_epoch_cursor` converts an epoch number to the type of its datetime or date partner before the comparison, so a mixed pair now orders. `test_promote_never_moves_last_value_backwards_across_an_epoch_and_a_string_cursor` covers both sides of the mix, a naive ISO string, and a date field. Its four cases fail on the previous comparison code.

File: products/warehouse_sources/backend/models/external_data_schema.py [hunk 2]
```diff
@@ -126,6 +126,14 @@ def _schema_ids_with_running_jobs(schema_ids: list[uuid.UUID]) -> set[uuid.UUID]
     )
 
 
+# A staged incremental cursor belongs to one extract attempt, and the batch carrying it can land
+# after a later attempt has already staged its own. Keeping only the newest slot dropped the earlier
+# attempt's cursor, so that batch's completion promoted nothing. Displaced slots are parked here and
+# stay promotable until their own batch lands. Bounded, because an attempt that never completes must
+# not grow the config without limit.
```
[the remark this answers is not in the document]
Comment (jakesciotto) [reply products/warehouse_sources/backend/models/external_data_schema.py]: Fixed in ec0f64e. `STAGED_CURSOR_PENDING_LIMIT` is now 15, which matches the production `MAX_RESUMABLE_SOURCE_RETRIES`, so no attempt of one run is evicted before its batch lands.

**Comment by jakesciotto** [comment]
**What happens next**

This PR stops a sync from losing its place when a long run retries. It is the first half of #83855.

1. After it ships, watch a newest-first table. Its watermark should be set after the next completed run, and each run should read less than before.
2. Second half: a table that cannot finish its first full read in one run should continue where it stopped, not start again from the newest row. We wait for step 1 to show if this is needed.
3. Small follow-ups, each its own PR: clear the staged cursor when someone changes the incremental field, and make the "never move the watermark backwards" check work for ObjectID and mixed date formats.
4. Tables that already have duplicate rows need one reset after the watermark is stable.

**Comment by danielcarletti** [comment]
Pushed two commits here after reviewing this next to #98762: a merge of master, and a fix commit. Nothing rebased or force-pushed, so revert freely if any of it is unwanted.

- **Blocker:** `STAGED_CURSOR_PENDING_LIMIT` was 15, pinned to the retry cap when the branch was cut. #97803 raised `MAX_RESUMABLE_SOURCE_RETRIES` to 20 on master, so a 20-attempt run could park 19 cursors and the trim would evict a live one. The production cap now lives in `data_imports/retry_limits.py`, read by both the activity and the model, so they cannot drift. The guard test compared against the DEBUG value (3) and could not catch this, so it is gone.
- **ObjectID ordering:** `_compare_incremental_values` returned None for strings, so ObjectID cursors fell to "newest promotion wins". With out-of-order promotion now possible, a late older attempt would move a Mongo watermark backwards. Equal-length ObjectIDs order as strings (leading timestamp). One parameterized case covers it; the "cannot be ordered" test now uses a null field type.
- **Retry on a dropped connection:** `stage_incremental_field_value` wraps its `update_sync_type_config_keys` call in `retry_on_db_connection_drop` like `record_partition_measurement`. `promote` stays unwrapped since it runs inside the completion transaction.
- **Unreachable same-run merge:** a run is live or parked, never both, because only another run's staging parks it and its own staging moves it back, all under the lock. Dropped the dedupe in `_park_displaced_staged_cursor`, the `{**parked, **live}` merge, and the two tests for that state. The PR body still advertises the merge; worth a one-line edit.
- Nits: trimmed the constant's comment, `direction` renamed to `kind`, `staged_cursor_promoted`/`staged_cursor_missing` now carry `team_id` and `external_data_job_id`, and the two "displaced cursor" stage tests are folded into one.

Verified locally: `TestStagedIncrementalCursor`, `TestStagedIncrementalCursorStaleWriters`, and `load/test_processor.py` (40 passed), ruff, ty, and repo-wide mypy.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: products/warehouse_sources/backend/models/external_data_schema.py [hunk 3]
```diff
@@ -911,31 +919,68 @@ def stamp_last_repartition_at(self) -> None:
...
+                )
+            if "earliest_value" in staged:
+                _advance_promoted_cursor(
+                    config, "incremental_field_earliest_value", staged["earliest_value"], "earliest", field_type
+                )
+            if live is not None:
+                config.pop("incremental_staged", None)
+
+        self.sync_type_config = update_sync_type_config_keys(self.id, self.team_id, mutate=mutate)
+        return found
```
[the remark this answers is not in the document]
Comment (danielcarletti) [reply products/warehouse_sources/backend/models/external_data_schema.py:983]: AI reply: Leaving `promote` unwrapped on purpose. It runs inside the load consumer's completion transaction (see the comment on `_promote_staged_cursor`), so a retry after a dropped connection would execute on a broken outer transaction. The right recovery there is the one that already exists: the failure rolls the completion back and the batch retries, which runs promotion again. `stage` is different: it runs on its own connection per batch, so it gets the one retry the sibling `record_partition_measurement` has.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by danielcarletti** [comment]
AI reply to CodeRabbit's grouped note on `_advance_promoted_cursor` (keep the current `earliest_value` when a pair cannot be ordered): not applying it.

Both promoted values come from batches that landed, so the loaded history reaches down to the lower of the two. Neither rule can pick that value without an ordering, so each of them can cost a bounded re-read after an out-of-order landing. The difference is the in-order case, which is the normal one. With "keep current", the first promotion sets `earliest_value` and every later run keeps it, so a newest-first source on an unordered cursor type never records backward progress and re-reads the same window on every run. With "newest wins" it follows the latest completed run, which is the behavior every cursor type had before this PR. ObjectID now orders, so the fallback only reaches a null field type, a naive against an aware datetime, and a millisecond epoch, which the PR body lists as follow-ups.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by danielcarletti** [comment]
AI reply to @jakesciotto's process question. The plan is sound, with four notes:

1. **Watching a newest-first table after ship.** The signal is `incremental_field_last_value` appearing on the schema after the next run whose final batch lands, and the `staged_cursor_promoted` log line for that run uuid. Each following run should send a `> last_value` window and count fewer rows. If a run still completes with `staged_cursor_missing`, the attempt that produced the final batch had its cursor evicted or never staged, and that is the case to look at first.
2. **Second half, continuing a first pass across runs.** Agreed to wait for step 1. The promoted `earliest_value` after this fix is the evidence: if it stops moving while the table is still short of the oldest record, #83855's second remedy is needed.
3. **Follow-ups.** Clearing the staged and parked cursors on an incremental field change is still open. The "never move backwards" gap is now closed in this PR for ObjectID (string order, same length) and for mixed epoch and ISO datetimes (`_align_epoch_cursor`). What remains unordered is a null field type, a naive against an aware datetime, and a millisecond epoch.
4. **One reset for tables that already hold duplicates.** Right, and only append tables need it, since merge dedupes on the primary key. The reset path clears the parked list too, so do it after the watermark has been promoted at least once, or the reset run starts from scratch again.

One more thing worth knowing: #97803 raised the resumable retry cap to 20 on master, which is why the parked-list bound now derives from a shared constant instead of a literal.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by danielcarletti — APPROVED** [review]

**Comment by danielcarletti** [comment]
Pushed 2401be3 after a second review pass: `_finalize` now stages the desc watermark before it sends the final-batch notification. The load consumer promotes the staged slot when that batch completes, so staging after the send left a `last_value` slot nothing promoted whenever the consumer won the race. This was the first item under "Not in this PR"; staging is promote-gated, so ordering it first has no downside. `TestFinalizeStagesTheWatermarkFirst` pins the order.

Also edited the description to match the branch: promotion no longer merges a same-run pair, the parked-list bound derives from the shared production cap, ObjectID now orders, and the resumed-run limitation (the final attempt stages only its own maximum) is listed under "Not in this PR" with its follow-up.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
